src/devices/InlineDevice.js

```diff
@@ -0,0 +1,257 @@
+ * session creation is possible on the XRDevice before
+ * the XRSession can be created.
+ */
+let SESSION_ID = 0;
+class Session {
+  constructor(mode, enabledFeatures) {
+    this.mode = mode;
+    this.enabledFeatures = enabledFeatures;
+    this.ended = null;
+    this.baseLayer = null;
+    this.id = ++SESSION_ID;
+  }
+};
+
+/**
+ * An XRDevice which only supports sensorless inline sessions, used as a
+ * fallback when no other type of XRDevice is available as a way to satisfy the
+ * spec requirement that inline sessions are always supported.
+ */
+export default class InlineDevice extends XRDevice {
+  /**
+   * Constructs an inline-only XRDevice
+   */
+  constructor(global) {
+    super(global);
+
+    this.sessions = new Map();
+    this.projectionMatrix = mat4.create();
+    this.identityMatrix = mat4.create();
+  }
+
+  /**
+   * Called when a XRSession has a `baseLayer` property set.
+   *
+   * @param {number} sessionId
+   * @param {XRWebGLLayer} layer
+   */
+  onBaseLayerSet(sessionId, layer) {
+    const session = this.sessions.get(sessionId);
+    session.baseLayer = layer;
+  }
+
+  /**
+   * Returns true if the requested mode is inline
+   *
+   * @param {XRSessionMode} mode
+   * @return {boolean}
+   */
+  isSessionSupported(mode) {
+    return mode == 'inline';
+  }
+
+  /**
+   * @param {string} featureDescriptor
+   * @return {boolean}
+   */
+  isFeatureSupported(featureDescriptor) {
+    switch(featureDescriptor) {
+      // Only viewer reference spaces are supported
+      case 'viewer': return true;
+      default: return false;
+    }
+  }
+
+  /**
+   * Returns a promise of a session ID if creating a session is successful.
+   *
+   * @param {XRSessionMode} mode
+   * @param {Set<string>} enabledFeatures
+   * @return {Promise<number>}
+   */
+  async requestSession(mode, enabledFeatures) {
+    if (!this.isSessionSupported(mode)) {
+      return Promise.reject();
+    }
+
+    const session = new Session(mode, enabledFeatures);
+
+    this.sessions.set(session.id, session);
+
+    return Promise.resolve(session.id);
+  }
+
+  /**
+   * @return {Function}
+   */
+  requestAnimationFrame(callback) {
+    return window.requestAnimationFrame(callback);
+  }
+
+  /**
+   * @param {number} handle
+   */
+  cancelAnimationFrame(handle) {
+    window.cancelAnimationFrame(handle);
+  }
+
+  onFrameStart(sessionId, renderState) {
+    // @TODO Our test environment doesn't have the canvas package for now,
+    // but this could be something we add to the tests.
+    if (TEST_ENV) {
+      return;
+    }
+
+    const session = this.sessions.get(sessionId);
+
+    // If the session is inline make sure the projection matrix matches the 
+    // aspect ratio of the underlying WebGL canvas.
+    if (session.baseLayer) {
+      const canvas = session.baseLayer.context.canvas;
+      // Update the projection matrix.
+      mat4.perspective(this.projectionMatrix, renderState.inlineVerticalFieldOfView,
```

Comment:
Realistically these values won't change often if at all -- could cache in the future after checking no changes in canvas aspect, near/far/fov values

Reply:
I like this idea, but I'd also like to do it in a more general way. I have an idea of how to approach it as a follow up PR.